src/borg/item.pyx

```diff
@@ -359,60 +377,127 @@ class ManifestItem(PropDict):
-        if item2.get('deleted') and not item1.get('deleted'):
+    def _dir_string(self):
+        if self.item2.get('deleted') and not self.item1.get('deleted'):
             return 'removed directory'
-        if item1.get('deleted') and not item2.get('deleted'):
+        if self.item1.get('deleted') and not self.item2.get('deleted'):
             return 'added directory'
+        return ''
 
-    def compare_owner(item1, item2):
-        user1, group1 = get_owner(item1)
-        user2, group2 = get_owner(item2)
-        if user1 != user2 or group1 != group2:
-            return '[{}:{} -> {}:{}]'.format(user1, group1, user2, group2)
+    def _owner_string(self):
+        if self._numeric_owner:
+             if self.item1.get('uid') != self.item2.get('uid') or self.item1.get('gid') != self.item2.get('gid'):
+                 return '[{}:{} -> {}:{}]'.format(self.item1.get('uid'), self.item1.get('gid'),
+                                                  self.item2.get('uid'), self.item2.get('gid'))
+        else:
+            if self.item1.get('user') != self.item2.get('user') or self.item1.get('group') != self.item2.get('group'):
+                 return '[{}:{} -> {}:{}]'.format(self.item1.get('user'), self.item1.get('group'),
+                                                  self.item2.get('user'), self.item2.get('group'))
+        return ''
 
-    def compare_mode(item1, item2):
-        if item1.mode != item2.mode:
-            return '[{} -> {}]'.format(get_mode(item1), get_mode(item2))
+    def _mode_string(self):
+        if 'mode' in self.item1 and 'mode' in self.item2 and self.item1.mode != self.item2.mode:
+            return '[{} -> {}]'.format(stat.filemode(self.item1.mode), stat.filemode(self.item2.mode))
+        return ''
 
-    def contents_changed(item1, item2):
+    def _content_equal(self, chunk_iterator1, chunk_iterator2, can_compare_chunk_ids):
         if can_compare_chunk_ids:
-            return item1.chunks != item2.chunks
-        if sum_chunk_size(item1) != sum_chunk_size(item2):
-            return True
-        chunk_ids1 = [c.id for c in item1.chunks]
-        chunk_ids2 = [c.id for c in item2.chunks]
-        return not fetch_and_compare_chunks(chunk_ids1, chunk_ids2, archive1, archive2)
+            return self.item1.chunks == self.item2.chunks
+        if self.item1.get_size() != self.item2.get_size():
+            return False
+        return ItemDiff._chunk_contents_equal(chunk_iterator1, chunk_iterator2)
 
     @staticmethod
-    def compare_chunk_contents(chunks1, chunks2):
-        """Compare two chunk iterators (like returned by :meth:`.DownloadPipeline.fetch_many`)"""
+    def _chunk_contents_equal(chunks1, chunks2):
+        """
+        Compare chunk contents and return True if they are identical.
+
+        The chunks must be given as chunk iterators (like returned by :meth:`.DownloadPipeline.fetch_many`).
+        """
+
         end = object()
         alen = ai = 0
         blen = bi = 0
```

Comment:
It's of course possible to have code coverage for C, its just not configured in this repository.

Cython on the other hand is more difficult because you'd need something like sourcemaps and I never heard of any tooling like that for Cython.